### PR TITLE
TMGLViewController: Initial Commit.

### PR DESCRIPTION
**Pull request by hagaiw**
TMViewController: Initial Commit.

Last PR includes all view controllers of ToneMapper.

### CODE REVIEW

File: ToneMapper/TMGLViewController.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+};
+
+/// Loads the given \c image into the workspace.
+- (void)loadTextureFromImage:(UIImage *)image;
+
+/// Saves the image currently displayed to disk.
+- (void)saveProcessedTexture;
+
+/// Applies the given translation to the displayed image.
+/// \c movementEnded indicates whether the current translation process ended or is still ongoing.
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.h]: Update doc.

File: ToneMapper/TMGLViewController.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+- (void)moveTextureWithTranslation:(CGPoint)translation gestureState:(TMGestureState)gestureState;
+
+/// Applies a scale transformation with the given \c scale to the x and y axis.
+/// \c zoomLocation is the centroid around which the zoom operation should occur.
+/// \c zoomEnded indicates whether the zoom operation has ended or is still ongoing.
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.h]: same

File: ToneMapper/TMGLViewController.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+- (void)zoomImageByScale:(GLfloat)scale zoomLocation:(CGPoint)zoomLocation
+               gestureState:(TMGestureState)gestureState;
+
+/// The maximum dimension size of for an image to be loaded as a texture.
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.h]: Doc

File: ToneMapper/TMGLViewController.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+- (NSUInteger)maxTextureSize;
+
+/// Applies the given toneMatrix to the image's color values.
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.h]: Doc

File: ToneMapper/TMGLViewController.m [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+#import "TMVector2DUniform.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface TMGLViewController () <GLKViewDelegate>
+
+/// Display that presents the processed texture in a scrollable manner.
+@property (strong, nonatomic) TMTextureDisplay *display;
+
+/// \c TMTexture representation of the image after applying the most recent filter.
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.m]: Correct doc?

File: ToneMapper/TMGLViewController.m [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+@property (strong, nonatomic) TMTexture *inputTexture;
+
+/// \c TMTexture representation of the image after changing the most recent filter.
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.m]: "after changing"

File: ToneMapper/TMGLViewController.m [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+@property (strong, nonatomic) TMTexture *processedTexture;
+
+/// Preprocessed \c TMTexture created by applying a bilateral filter with a strong blur to the
+/// loaded image.
+//@property (strong, nonatomic) TMTexture *strongBlurTexture;
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.m]: Remove comment

File: ToneMapper/TMGLViewController.m [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+/// Display that presents the processed texture in a scrollable manner.
+@property (strong, nonatomic) TMTextureDisplay *display;
+
+/// \c TMTexture representation of the image after applying the most recent filter.
+@property (strong, nonatomic) TMTexture *inputTexture;
+
+/// \c TMTexture representation of the image after changing the most recent filter.
+@property (strong, nonatomic) TMTexture *processedTexture;
+
+/// Preprocessed \c TMTexture created by applying a bilateral filter with a strong blur to the
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.m]: Update doc.

File: ToneMapper/TMGLViewController.m [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+/// loaded image.
+//@property (strong, nonatomic) TMTexture *strongBlurTexture;
+@property (strong, nonatomic) TMTextureUniform *strongBlurTextureUniform;
+
+/// Preprocessed \c TMTexture created by applying a bilateral filter with a weak blur to the
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.m]: Doc

File: ToneMapper/TMGLViewController.m [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+/// Processor for global tone adjustments.
+@property (strong, nonatomic) id<TMProcessor> globalTonesProcessor;
+
+/// Processor for global tone adjustments.
+@property (strong, nonatomic) id<TMProcessor> bilateralProcessor;
+
+/// Processor to use when processing the next frame.
+@property (strong, nonatomic) id<TMProcessor> currentProcessor;
+
+@property (strong, nonatomic) TMMatrixUniform *yFlipProjection;
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.m]: Doc

File: ToneMapper/TMGLViewController.m [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+                                                                initWithGLKView:self.glkView]
+                  program:[self.programFactory passThroughWithProjection]
+                  geometry:[self.geometryFactory quadGeometry]];
+}
+
+- (void)setupGLKView {
+  if (self.glkView) {
+    [self.glkView removeFromSuperview];
+  }
+  self.glkView = [GLKView new];
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.m]: `alloc init`

File: ToneMapper/TMGLViewController.m [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+- (void)initProperties {
+  self.programFactory = [[TMProgramFactory alloc] init];
+  self.geometryFactory = [[TMGeometryFactory alloc] init];
+  self.processorFactory = [[TMTextureProcessorFactory alloc] init];
+  self.positionFactory = [[TMPositionFactory alloc] init];
+  self.projectionFactory = [[TMProjectionFactory alloc] init];
+  self.toneAdjustmentGenerator = [[TMToneAdjustmentGenerator alloc] init];
+  self.textureNeedsProcessing = true;
+  self.yFlipProjection = [[TMMatrixUniform alloc] initWithMatrix:[self.projectionFactory
+                                                               verticalMirrorProjection]
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.m]: Indentation

File: ToneMapper/TMGLViewController.m [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+                                                      uniform:kProjectionUniform];
+}
+
+- (void)initProcessors {
+  TMTextureProgram *mixerProgram = [self.programFactory textureMixingProgram];
+  self.localContrastProcessor = [self.processorFactory processorWithProgram:mixerProgram];
+
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.m]: Redundant line break.

File: ToneMapper/TMGLViewController.m [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+- (void)initProcessors {
+  TMTextureProgram *mixerProgram = [self.programFactory textureMixingProgram];
+  self.localContrastProcessor = [self.processorFactory processorWithProgram:mixerProgram];
+  
+  TMTextureProgram *horizontalBilateral = [self.programFactory bilateralHorizontalFilterProgram];
+  TMTextureProgram *verticalBilateral = [self.programFactory bilateralVerticalFilterProgram];
+  self.bilateralProcessor = [self.processorFactory processorWithPrograms:@[horizontalBilateral,
+                                                                           verticalBilateral]];
+
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.m]: same

File: ToneMapper/TMGLViewController.m [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+#pragma mark -
+
+- (void)moveTextureWithTranslation:(CGPoint)translation
+                      gestureState:(TMGestureState)gestureState {
+  TMScaledPosition *positionDelta = [self.positionFactory positionWithTranslation:translation];
+  [self updateTexturePositionWithPosition:positionDelta gestureState:gestureState];
+}
+
+- (void)zoomImageByScale:(GLfloat)scale zoomLocation:(CGPoint)zoomLocation
+               gestureState:(TMGestureState)gestureState {
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.m]: Indentation

File: ToneMapper/TMGLViewController.m [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+
+#pragma mark -
+#pragma mark Tone Adjustments
+#pragma mark -
+
+- (void)setLocalContrastWithMediumWeight:(GLfloat)mediumWeight fineWeight:(GLfloat)fineWeight {
+  TMScalarUniform *mediumWeightUniform = [[TMScalarUniform alloc] initWithName:kMediumWeight
+                                                                        scalar:mediumWeight];
+  TMScalarUniform *fineWeightUniform = [[TMScalarUniform alloc] initWithName:kFineWeight
+                                                                     scalar:fineWeight];
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.m]: same

File: ToneMapper/TMGLViewController.m [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+  toneMatrix = GLKMatrix4Multiply([self.toneAdjustmentGenerator
+                                   toneMatrixForTemperature:temperature], toneMatrix);
+  TMMatrixUniform *toneMatrixUniform = [[TMMatrixUniform alloc] initWithMatrix:toneMatrix
+                                                                       uniform:kToneAdjustment];
+  self.currentUniforms = @[toneMatrixUniform];
+  self.currentProcessor = self.globalTonesProcessor;
+}
+
+- (TMTexture *)texture:(TMTexture *)texture
+      FromBilateralWithNumberOfPasses:(NSUInteger)numberOfPasses {
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.m]: `From` -> `from`

File: ToneMapper/TMGLViewController.m [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+  GLint maxTextureSize;
+  glGetIntegerv(GL_MAX_TEXTURE_SIZE, &maxTextureSize);
+  return (NSUInteger)maxTextureSize;
+}
+
+- (void)loadTextureFromImage:(UIImage *)image {
+  self.inputTexture = [[TMTexture alloc] initWithImage:image];
+  self.processedTexture = self.inputTexture;
+  [self setupTextureWorkspace];
+
```
Comment (CodeBaseCamp) [inline ToneMapper/TMGLViewController.m]: Remove line break.

File: ToneMapper/TMViewController.h [hunk 19]
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) 2015 Lightricks. All rights reserved.
+// Created by Hagai Weinfeld.
+
+#import <UIKit/UIKit.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/// Main view-controller, handles user-interaction and app interface.
+@interface TMViewController : UIViewController <UINavigationControllerDelegate, UIImagePickerControllerDelegate>
```
Comment (CodeBaseCamp) [inline ToneMapper/TMViewController.h]: 100

File: ToneMapper/TMViewController.m [hunk 20]
```diff
@@ -0,0 +1,302 @@
...
+#pragma mark UIButton
+#pragma mark -
+
+- (IBAction)saveImage:(UIButton *)sender {
+  [self.openGLVC saveProcessedTexture];
+}
+
+- (IBAction)moveImage:(UIPanGestureRecognizer *)sender {
+  GLfloat dx = [sender translationInView:self.openGLVC.view].x /
+  self.openGLVC.view.frame.size.width;
```
Comment (CodeBaseCamp) [inline ToneMapper/TMViewController.m]: Indentation

File: ToneMapper/TMViewController.m [hunk 21]
```diff
@@ -0,0 +1,302 @@
...
+  GLfloat dy = [sender translationInView:self.openGLVC.view].y /
+  self.openGLVC.view.frame.size.height;
```
Comment (CodeBaseCamp) [inline ToneMapper/TMViewController.m]: Indentation

File: ToneMapper/TMViewController.m [hunk 22]
```diff
@@ -0,0 +1,302 @@
...
+  [self.openGLVC moveTextureWithTranslation:CGPointMake(dx * [UIScreen mainScreen].scale,
+                                                        -dy * [UIScreen mainScreen].scale)
+                               gestureState:([sender state] == UIGestureRecognizerStateEnded ?
+                                             TMGestureStateEnded : TMGestureStateOngoing) ];
```
Comment (CodeBaseCamp) [inline ToneMapper/TMViewController.m]: `) ];` -> `)];`

File: ToneMapper/TMViewController.m [hunk 23]
```diff
@@ -0,0 +1,302 @@
...
+}
+
+- (IBAction)zoomImage:(UIPinchGestureRecognizer *)sender {
+  CGPoint point = [sender locationInView:self.openGLVC.view];
+  
+  // Convert to world coordinates.
+  CGFloat pinchX = (2*point.x - (self.openGLVC.view.frame.size.width) ) /
+  self.openGLVC.view.frame.size.width;
```
Comment (CodeBaseCamp) [inline ToneMapper/TMViewController.m]: Indentation

File: ToneMapper/TMViewController.m [hunk 24]
```diff
@@ -0,0 +1,302 @@
...
+  CGFloat pinchY = (2*point.y - (self.openGLVC.view.frame.size.height) ) /
+  self.openGLVC.view.frame.size.height;
```
Comment (CodeBaseCamp) [inline ToneMapper/TMViewController.m]: Indentation

File: ToneMapper/TMViewController.m [hunk 25]
```diff
@@ -0,0 +1,302 @@
...
+  [self.openGLVC zoomImageByScale:[sender scale] zoomLocation:CGPointMake(pinchX, pinchY)
+                     gestureState:([sender state] == UIGestureRecognizerStateEnded ?
+                                   TMGestureStateEnded : TMGestureStateOngoing) ];
+}
+
+- (IBAction)pickImage:(UIButton *)sender {
+  UIImagePickerController *mediaUI = [UIImagePickerController new];
```
Comment (CodeBaseCamp) [inline ToneMapper/TMViewController.m]: Find better name for `mediaUI`.

Comment (CodeBaseCamp) [inline ToneMapper/TMViewController.m]: `alloc init`

File: ToneMapper/TMViewController.m [hunk 26]
```diff
@@ -0,0 +1,302 @@
...
+        self.mediumContrastValue = sender.value;
+        break;
+      case FineContrastType:
+        self.fineContrastValue = sender.value;
+        break;
+      default:
+        break;
+    }
+    [self.openGLVC setLocalContrastWithMediumWeight:self.mediumContrastValue
+                                           fineWeight:self.fineContrastValue];
```
Comment (CodeBaseCamp) [inline ToneMapper/TMViewController.m]: Indentation

File: ToneMapper/TMViewController.m [hunk 27]
```diff
@@ -0,0 +1,302 @@
...
+- (IBAction)sliderMoved:(UISlider *)sender {
+  if (self.localContrast.selectedSegmentIndex != kDeselectedValue) {
+    switch (self.currentLocalContrastType) {
+      case MediumContrastType:
+        self.mediumContrastValue = sender.value;
+        break;
+      case FineContrastType:
+        self.fineContrastValue = sender.value;
+        break;
+      default:
```
Comment (CodeBaseCamp) [inline ToneMapper/TMViewController.m]: Remove default case and handle `NoneContrastType` explicitly.

File: ToneMapper/TMViewController.m [hunk 28]
```diff
@@ -0,0 +1,302 @@
...
+};
+
+/// Types of tone-adjustments.
+typedef NS_ENUM(NSInteger, ToneAdjustment) {
+  ToneAdjustmentBrightness,
+  ToneAdjustmentContrast,
+  ToneAdjustmentSaturation,
+  ToneAdjustmentTint,
+  ToneAdjustmentTemperature,
+  ToneAdjustmentNone
```
Comment (CodeBaseCamp) [inline ToneMapper/TMViewController.m]: Required?

File: ToneMapper/TMViewController.m [hunk 29]
```diff
@@ -0,0 +1,302 @@
...
+      case ToneAdjustmentSaturation:
+        self.saturationValue = sender.value;
+        break;
+      case ToneAdjustmentTint:
+        self.tintValue = sender.value;
+        break;
+      case ToneAdjustmentTemperature:
+        self.temperatureValue = sender.value;
+        break;
+      default:
```
Comment (CodeBaseCamp) [inline ToneMapper/TMViewController.m]: same

File: ToneMapper/TMViewController.m [hunk 30]
```diff
@@ -0,0 +1,302 @@
...
+  UIImage *normalizedImage = UIGraphicsGetImageFromCurrentImageContext();
+  UIGraphicsEndImageContext();
+  return normalizedImage;
+}
+
+#pragma mark -
+#pragma mark UIImagePickerController Delegate
+#pragma mark -
+
+- (void) imagePickerController: (UIImagePickerController *) picker
```
Comment (CodeBaseCamp) [inline ToneMapper/TMViewController.m]: White spaces

File: ToneMapper/TMViewController.m [hunk 31]
```diff
@@ -0,0 +1,302 @@
...
+ didFinishPickingMediaWithInfo: (NSDictionary *) info {
```
Comment (CodeBaseCamp) [inline ToneMapper/TMViewController.m]: Indentation

**Comment by CodeBaseCamp** [comment]
Overall, the design looks pretty decent. However, I would suggest to extract independent logic (image loading, storing, translation, etc.) from `TMGLViewController` and have it be responsible only for processing and drawing. The VC can receive the relevant information (such as position) from a suitable provider.

**Comment by CodeBaseCamp** [comment]
Round 1.

File: ToneMapper/TMGLViewController.h (hunk 1 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.h]: Done.

File: ToneMapper/TMGLViewController.h (hunk 2 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.h]: Done.

File: ToneMapper/TMGLViewController.h (hunk 3 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.h]: Done.

File: ToneMapper/TMGLViewController.h (hunk 4 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.h]: Done.

File: ToneMapper/TMGLViewController.m (hunk 6 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.m]: Done.

File: ToneMapper/TMGLViewController.m (hunk 7 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.m]: Done.

File: ToneMapper/TMGLViewController.m (hunk 8 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.m]: Done.

File: ToneMapper/TMGLViewController.m (hunk 9 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.m]: Done.

File: ToneMapper/TMGLViewController.m (hunk 10 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.m]: Done.

File: ToneMapper/TMGLViewController.m (hunk 11 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.m]: Done.

File: ToneMapper/TMGLViewController.m (hunk 12 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.m]: Done.

File: ToneMapper/TMGLViewController.m (hunk 13 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.m]: Done.

File: ToneMapper/TMGLViewController.m (hunk 14 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.m]: Done.

File: ToneMapper/TMGLViewController.m (hunk 15 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.m]: Done.

File: ToneMapper/TMGLViewController.m (hunk 16 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.m]: Done.

File: ToneMapper/TMGLViewController.m (hunk 17 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.m]: Done.

File: ToneMapper/TMGLViewController.m (hunk 18 above)
Comment (hagaiw) [inline ToneMapper/TMGLViewController.m]: Done.

File: ToneMapper/TMViewController.h (hunk 19 above)
Comment (hagaiw) [inline ToneMapper/TMViewController.h]: Done.

File: ToneMapper/TMViewController.m (hunk 20 above)
Comment (hagaiw) [inline ToneMapper/TMViewController.m]: Done.

File: ToneMapper/TMViewController.m (hunk 21 above)
Comment (hagaiw) [inline ToneMapper/TMViewController.m]: Done.

File: ToneMapper/TMViewController.m (hunk 22 above)
Comment (hagaiw) [inline ToneMapper/TMViewController.m]: Done.

File: ToneMapper/TMViewController.m (hunk 23 above)
Comment (hagaiw) [inline ToneMapper/TMViewController.m]: Done.

File: ToneMapper/TMViewController.m (hunk 24 above)
Comment (hagaiw) [inline ToneMapper/TMViewController.m]: Done.

File: ToneMapper/TMViewController.m (hunk 25 above)
Comment (hagaiw) [inline ToneMapper/TMViewController.m]: Done.

File: ToneMapper/TMViewController.m (hunk 26 above)
Comment (hagaiw) [inline ToneMapper/TMViewController.m]: Done.

File: ToneMapper/TMViewController.m (hunk 27 above)
Comment (hagaiw) [inline ToneMapper/TMViewController.m]: Done.

File: ToneMapper/TMViewController.m (hunk 28 above)
Comment (hagaiw) [inline ToneMapper/TMViewController.m]: Yes. Added `self.currentToneAdjustment = ToneAdjustmentNone;` to `- (void)resetAdjustments`

File: ToneMapper/TMViewController.m (hunk 29 above)
Comment (hagaiw) [inline ToneMapper/TMViewController.m]: Done.

File: ToneMapper/TMViewController.m (hunk 30 above)
Comment (hagaiw) [inline ToneMapper/TMViewController.m]: Done.

File: ToneMapper/TMViewController.m (hunk 31 above)
Comment (hagaiw) [inline ToneMapper/TMViewController.m]: Done.

**Comment by hagaiw** [comment]
Round 1 fixes.
I've extracted the saving(storing) logic into a separate class - `TMTextureSaver`.
I left the rest of the "independent logic" in `TMGLViewController` because I felt like it's too interconnected in my current design.
Should I rewrite these parts or can I leave them the way they are for the sake of this exercise?

**Comment by CodeBaseCamp** [comment]
Round 2. I commented on commit by accident. :dizzy_face:

**Comment by CodeBaseCamp** [comment]
Regarding the extraction of logic not directly related to drawing: let's leave it as is for now. However, please be aware that your feeling that it is "too interconnected" alludes to the fact that the design is suboptimal. (please see also https://en.wikipedia.org/wiki/Coupling_(computer_programming) and https://en.wikipedia.org/wiki/Cohesion_(computer_science))

**Comment by hagaiw** [comment]
Round 2 fixes.
Also regarding my design, I know it's suboptimal :) I will look into those links though, thanks!

**Comment by CodeBaseCamp** [comment]
Squash.

**Comment by hagaiw** [comment]
Squashed.

**Pull request merged**